src/base/utils/tar.cpp

```diff
@@ -0,0 +1,230 @@
+                if (byte >= '0' && byte <= '7')
+                    break;
+            }
+
+            int fileSize = 0;
+            for (; i < data.size(); ++i) {
+                char byte = data.at(i);
+                if (byte < '0' || byte > '7') // skip trailing invalid bytes
+                    break;
+
+                fileSize = (8 * fileSize) + (byte - '0');
+            }
+
+            return fileSize;
+        }
+
+        // The various header fields that hold strings are of known max size.
+        // So in the case the string occupies their max size, the string will not
+        // be null terminated. It is pointless to have std::strlen() run through
+        // the whole buffer when we already know the upper limit. Also the buffer
+        // isn't guaranteed to contain a null byte and std::strlen() will have
+        // undefined behavior.
+        int customStrlen(const QByteArray &str) const
+        {
+            int sz = 0;
+            for (; sz < str.size(); ++sz) {
+                if (str.at(sz) == '\0')
+                    break;
+            }
+
+            return sz;
+        }
+
+        // Verified: When the checksum field gets verified
+        // Null: When all bytes are null. Used to indicate end-of-archive (EOF).
+        // Since both actions require running through the whole buffer, combine them to avoid running
+        // twice through the buffer.
+        void verifyAndCheckNull()
+        {
+            int calculatedChecksum = 0;
+            for (int i = 0; i < TAR_CHECKSUM_OFFSET; ++i)
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            calculatedChecksum += TAR_CHECKSUM_SIZE * ' '; // 8 ASCII spaces for the checksum field
+
+            for (int i = TAR_FILETYPE_OFFSET; i < m_header.size(); ++i)
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            const int checkSum = octalToDec(Utils::ByteArray::midView(m_header, TAR_CHECKSUM_OFFSET, TAR_CHECKSUM_SIZE));
+            m_null = (calculatedChecksum == TAR_CHECKSUM_SIZE * ' ') && (checkSum == 0);
+            m_verified = calculatedChecksum == checkSum;
+        }
+
+        bool m_null = true;
+        bool m_valid = false;
+        bool m_verified = false;
+
+        const QByteArray m_header;
+    };
+}
+
+
+QHash<QString, QByteArray> Utils::Tar::decompress(const QByteArray &data, bool *ok)
+{
+    if (ok)
+        *ok = false;
+    // Valid data data to us are at 1 header + payload
+    if (data.size() < TAR_HEADER_SIZE + 1)
+        return {};
+
+    int curPos = 0;
+    int dataSize = data.size();
+    QHash<QString, QByteArray> files;
+
+    while ((curPos < dataSize) && files.size() < 11) {
+        try {
+            TarEntry file(Utils::ByteArray::midView(data, curPos, TAR_HEADER_SIZE));
+            if (file.isNull()) // Possibly EOF
+                break;
+
+            if (!file.isValid()) {
+                LogMsg(QCoreApplication::translate("Utils::Tar",
```

Comment:
It's wrong. Here you just unpacking some tar archive (it's abstract utility function). You should perform any logging at the application level.